README.md
<div align="center">
  <a href="https://piotrmurach.github.io/tty" target="_blank"><img width="130" src="https://github.com/piotrmurach/tty/raw/master/images/tty.png" alt="tty logo" /></a>
</div>

# TTY::Exit [![Gitter](https://badges.gitter.im/Join%20Chat.svg)][gitter]

[![Gem Version](https://badge.fury.io/rb/tty-exit.svg)][gem]
[![Actions CI](https://github.com/piotrmurach/tty-exit/workflows/CI/badge.svg?branch=master)][gh_actions_ci]
[![Build status](https://ci.appveyor.com/api/projects/status/rtm3po27ouarfrhf?svg=true)][appveyor]
[![Code Climate](https://codeclimate.com/github/piotrmurach/tty-exit/badges/gpa.svg)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/github/piotrmurach/tty-exit/badge.svg)][coverage]
[![Inline docs](http://inch-ci.org/github/piotrmurach/tty-exit.svg?branch=master)][inchpages]

[gitter]: https://gitter.im/piotrmurach/tty
[gem]: http://badge.fury.io/rb/tty-exit
[gh_actions_ci]: https://github.com/piotrmurach/tty-exit/actions?query=workflow%3ACI
[appveyor]: https://ci.appveyor.com/project/piotrmurach/tty-exit
[codeclimate]: https://codeclimate.com/github/piotrmurach/tty-exit
[coverage]: https://coveralls.io/github/piotrmurach/tty-exit
[inchpages]: http://inch-ci.org/github/piotrmurach/tty-exit

> Terminal exit codes for humans and machines.

The goal of this library is to provide human friendly and standard way to use exit status codes in command line applications. Instead of saying `exit(64)`, you can say `exit_with(:usage_error)`. Both indicate a failure to the parent process but the `:usage_error` is so much nicer! Wouldn't you agree? That's why `tty-exit` gathers a list of all the most common exit codes as used by POSIX-compliant tools on different Unix systems for you to use.

The exit statuses range from 0 to 255 (inclusive). Any other exit status than 0 indicates a failure of some kind. The exit codes in the range 64-78 are adapted from the OpenBSD [sysexits.h](https://man.openbsd.org/sysexits.3). The codes between 125 and 128 are reserved for shell statuses as defined in [Advanced Bash Scripting Guide, Appendix E](http://tldp.org/LDP/abs/html/exitcodes.html). The codes in the 129-154 range correspond with the fatal signals as defined in [signal](https://man.openbsd.org/signal.3).

**TTY::Exit** provides independent terminal exit codes component for [TTY](https://github.com/piotrmurach/tty) toolkit.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'tty-exit'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install tty-exit

## Contents

* [1. Usage](#1-usage)
* [2. API](#2-api)
  * [2.1 exit_code](#21-exit_code)
  * [2.2 exit_message](#22-exit_message)
  * [2.3 exit_with](#23-exit_with)
  * [2.4 register_exit](#24-register_exit)
  * [2.5 exit_reserved?](#25-exit_reserved)
  * [2.6 exit_valid?](#26-exit_valid)
  * [2.7 exit_success?](#27-exit_success)
  * [2.8 exit_codes](#28-exit_codes)
  * [2.9 exit_messages](#29-exit_messages)

## 1. Usage

To exit from any program use `exit_with` method. Instead of a number, you can use a readable name for the exit status:

```ruby
TTY::Exit.exit_with(:usage_error)
```

The above will exit program immediately with an exit code indicating a failure:

```ruby
puts $?.exitstatus
# => 64
```

All the reserved exit statuses have a matching exit message. To display a default message, as a second argument to `exit_with` you can pass `:default` value:

```ruby
TTY::Exit.exit_with(:usage_error, :default)
```

That will produce the following user friendly message:

```ruby
# => "ERROR(64): Command line usage error"
```

The preferred way is to include **TTY::Exit** module in your code:

```ruby
class Command
  include TTY::Exit

  def execute
    exit_with(:config_error, :default)
  end
end
```

This will print an error message and return appropriate exit status:

```ruby
cmd = Command.new
cmd.execute
# => "ERROR(78): Configuration Error"
puts $?.exitstatus
# => 78
```

To see the full list of reserved exit codes go to [2.8 exit_codes](#28-exit-codes) section.

## 2. API

### 2.1 exit_code

There are many built-in exit codes that can be referenced using a name.

For example to return an exit code denoting success, you can use `:ok` or `:success`:

```ruby
TTY::Exit.exit_code(:ok) # => 0
TTY::Exit.exit_code(:success) # => 0
```

Any other exit status than 0 indicates a failure of some kind. For example, when a command cannot be found use `:not_found`:

```ruby
TTY::Exit.exit_code(:not_found)
# => 127
```

You can also use an exit code directly:

```ruby
TTY::Exit.exit_code(127)
# => 127
```

### 2.2 exit_message

One of the downsides of exit codes is that they are not very communicative to the user. Why not have both? An exit code and a user friendly message. That's what `exit_message` is for. All the reserved exit codes have corresponding user friendly messages.

For example, when returning exit code `64` for usage error:

```ruby
TTY::Exit.exit_message(:usage_error)
TTY::Exit.exit_message(64)
```

Will return:

```ruby
# => "ERROR(64): Command line usage error"
```

The default messages are used by the `exit_with` method and can be overwritten by a custom one.

### 2.3 exit_with

To exit program with an exit code use `exit_with`. This method accepts a name or a code for the exit status. 

```ruby
TTY::Exit.exit_with(:usage_error)
TTY::Exit.exit_with(64)
```

Both will produce the same outcome.

As a second argument you can specify a user friendly message to be printed to `stderr` before exit. To use predefined messages use `:default` as a value:

```ruby
TTY::Exit.exit_with(:usage_error, :default)
# => "ERROR(64): Command line usage error"
```

Optionally, you can provide a custom message to display to the user.

```ruby
TTY::Exit.exit_with(:usge_error, "Wrong arguments")
# => "Wrong arguments"
```

Finally, you can redirect output to a different stream using `:io` option. By default, message is printed to `stderr`:

```ruby
TTY::Exit.exit_with(:usage_error, io: $stdout)
```

Since `TTY::Exit` is a module, you can include it in your code to get access to all the methods:

```ruby
class Command
  include TTY::Exit

  def execute
    exit_with(:usage_error, :default)
  end
end
```

### 2.4 register_exit

If the provided exit codes don't match your needs, you can add your own using the `register_exit` method. 

For example, to register a custom exit with `:too_long` name and the status `7` that will notify user and programs about too many arguments do:

```ruby
class Command
  include TTY::Exit

  register_exit(:too_long, 7, "Argument list too long")

  def execute
    exit_with(:too_long, :default)
  end
end
```

Then when the command gets run:

```ruby
cmd = Command.new
cmd.execute
# =>
# ERROR(7): Argument list too long
```

### 2.5 exit_reserved?

To check if an exit code is already reserved by Unix system use `exit_reserved?`. This is useful in situations where you want to add it your command line application custom exit codes. The check accepts only integer numbers in range 0 to 255 inclusive:

```ruby
TTY::Exit.exit_reserved?(126) # => true
TTY::Exit.exit_reserved?(100) # => false
```

### 2.6 exit_valid?

The exit statuses range from 0 to 255 (inclusive). The `exit_valid?` method helps you check if an exit status is within the range or not.

```ruby
TTY::Exit.exit_valid?(11) # => true
TTY::Exit.exit_valid?(522) # => false
```

### 2.7 exit_success?

Any other exit status than 0 indicates a failure of some kind. The `exit_success?` is a more descriptive way to determine if a program succeeded or not.

```ruby
TTY::Exit.exit_success?(0) # => true
TTY::Exit.exit_success?(7) # => false
```

### 2.8 exit_codes

You can access all the predefined exit codes and their names with `exit_codes` method:

```ruby
TTY::Exit.exit_codes
# =>
# "{:ok=>0, :success=>0, :error=>1, :shell_misuse=>2, :usage_error=>64, :data_error=>65, ... }"
```

### 2.9 exit_messages

To see what default messages are for the predefined exit codes use `exit_messages`:

```ruby
TTY::Exit.exit_messages
# =>
# "{0=>"Successful termination", 1=>"An error occurred", 2=>"Misuse of shell builtins", 64=>"Command line usage error", ... }"
```

The exit statuses range from 0 to 255 (inclusive). Any other exit status than 0 indicates a failure of some kind. The exit codes in the range 64-78 are adapted from the OpenBSD [sysexits.h](https://man.openbsd.org/sysexits.3). The codes between 125 and 128 are reserved for shell statuses as defined in [Advanced Bash Scripting Guide, Appendix E](http://tldp.org/LDP/abs/html/exitcodes.html). The codes in the 129-154 range correspond with the fatal signals as defined in [signal](https://man.openbsd.org/signal.3).


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/piotrmurach/tty-exit. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/piotrmurach/tty-exit/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the TTY::Exit project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/piotrmurach/tty-exit/blob/master/CODE_OF_CONDUCT.md).

## Copyright

Copyright (c) 2020 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
# Change log

## [v0.1.0] - 2020-02-01

* Initial implementation and release

[v0.1.0]: https://github.com/piotrmurach/tty-exit/compare/204b8ac...v0.1.0


spec/spec_helper.rb
if ENV["COVERAGE"] == "true"
  require "simplecov"
  require "coveralls"

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ])

  SimpleCov.start do
    command_name "spec"
    add_filter "spec"
  end
end

require "bundler/setup"
require "tty/exit"

RSpec.configure do |config|
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/unit/exit_code_spec.rb
# frozen_string_literal: true

RSpec.describe TTY::Exit, "#exit_code" do
  CODES = {
    ok: 0,
    success: 0,
    error: 1,
    shell_misuse: 2,

    usage_error: 64,
    data_err: 65,
    data_error: 65,
    no_input: 66,
    no_user: 67,
    no_host: 68,
    service_unavailable: 69,
    software_error: 70,
    system_error: 71,
    system_file_missing: 72,
    cant_create: 73,
    io_error: 74,
    io_err: 74,
    temp_fail: 75,
    protocol: 76,
    no_perm: 77,
    no_permission: 77,
    config_error: 78,
    configuration_error: 78,

    cannot_execute: 126,
    not_found: 127,
    command_not_found: 127,
    invalid_argument: 128,

    hangup: 129,
    interrupt: 130,
    quit: 131,
    illegal_instruction: 132,
    trace_trap: 133,
    abort: 134,
    kill: 137,
    bus_error: 138,
    memory_error: 139,
    segmentation_fault: 139,
    pipe: 141,
    alarm: 142,
    user1: 158,
    user2: 159
  }.freeze

  CODES.map do |name, code|
    it "converts #{name.inspect} to #{code.inspect} exit code" do
      expect(TTY::Exit.exit_code(name)).to eq(code)
    end
  end

  it "provides a mapping between names and codes" do
    expect(TTY::Exit.exit_codes.values).to eq(CODES.values)
  end

  it "accepts name as a string" do
    expect(TTY::Exit.exit_code("data_error")).to eq(65)
  end

  it "accepts a code directly" do
    expect(TTY::Exit.exit_code(65)).to eq(65)
  end

  it "returns 0 code when none provided" do
    expect(TTY::Exit.exit_code).to eq(0)
  end

  it "uses exit status constant" do
    expect(TTY::Exit.exit_code).to eq(TTY::Exit::Code::SUCCESS)
  end

  it "fails to find a code for an unknown name" do
    expect {
      TTY::Exit.exit_code(:unknown)
    }.to raise_error(TTY::Exit::Error, "Name 'unknown' isn't recognized.")
  end

  it "raises when code is of unknown type" do
    expect {
      TTY::Exit.exit_code(Object.new)
    }.to raise_error(TTY::Exit::Error, "Provide a name or a number as an exit code")
  end

  it "raises when code is outside of the allowed range of (0 - 255)" do
    expect {
      TTY::Exit.exit_code(666)
    }.to raise_error(TTY::Exit::Error, "Provided code outside of the range (0 - 255)")
  end
end


spec/unit/exit_with_spec.rb
# frozen_string_literal: true

RSpec.describe TTY::Exit, "#exit_with" do
  it "exposes exit_with when including a module" do
    stub_const("Command", Class.new do
      include TTY::Exit

      def execute
        exit_with(:usage_error)
      end
    end)

    cmd = Command.new

    expect {
      cmd.execute
    }.to raise_error(SystemExit) do |err|
      expect(err.status).to eq(64)
      message = RSpec::Support::Ruby.truffleruby? ? "SystemExit" : "exit"
      expect(err.message).to eq(message)
    end
  end

  it "prints a custom message with exit code when including a module" do
    output = StringIO.new

    stub_const("Command", Class.new do
      include TTY::Exit

      def initialize(io)
        @io = io
      end

      def execute
        exit_with(:usage_error, "Wrong arguments", io: @io)
      end
    end)

    cmd = Command.new(output)
    expect { cmd.execute }.to raise_error(SystemExit)
    expect(output.string).to eq("Wrong arguments")
  end

  it "prints a predefined message with exit code when including a module" do
    output = StringIO.new

    stub_const("Command", Class.new do
      include TTY::Exit

      def initialize(io)
        @io = io
      end

      def execute
        exit_with(:usage_error, :default, io: @io)
      end
    end)

    cmd = Command.new(output)
    expect { cmd.execute }.to raise_error(SystemExit)

    expect(output.string).to eq("ERROR(64): Command line usage error")
  end
end


lib/tty/exit.rb
# frozen_string_literal: true

require_relative "exit/code"
require_relative "exit/registry"
require_relative "exit/version"

module TTY
  # Terminal exit codes for humans and machines
  module Exit
    extend self

    Error = Class.new(StandardError)

    # @api private
    def self.included(base)
      base.instance_eval do
        def register_exit(*args)
          Registry.register_exit(*args)
        end
      end
    end

    NAME_TO_EXIT_CODE = {
      ok: Code::SUCCESS,
      success: Code::SUCCESS,
      error: Code::ERROR,
      shell_misuse: Code::SHELL_MISUSE,

      usage_error: Code::USAGE_ERROR,
      data_err: Code::DATA_ERROR,
      data_error: Code::DATA_ERROR,
      no_input: Code::NO_INPUT,
      no_user: Code::NO_USER,
      no_host: Code::NO_HOST,
      service_unavailable: Code::SERVICE_UNAVAILABLE,
      software_error: Code::SOFTWARE_ERROR,
      system_error: Code::SYSTEM_ERROR,
      system_file_missing: Code::SYSTEM_FILE_MISSING,
      cant_create: Code::CANT_CREATE,
      io_error: Code::IO_ERROR,
      io_err: Code::IO_ERROR,
      temp_fail: Code::TEMP_FAIL,
      protocol: Code::PROTOCOL,
      no_perm: Code::NO_PERM,
      no_permission: Code::NO_PERM,
      config_error: Code::CONFIG_ERROR,
      configuration_error: Code::CONFIG_ERROR,

      cannot_execute: Code::CANNOT_EXECUTE,
      not_found: Code::COMMAND_NOT_FOUND,
      command_not_found: Code::COMMAND_NOT_FOUND,
      invalid_argument: Code::INVALID_ARGUMENT,

      hangup: Code::HANGUP,
      interrupt: Code::INTERRUPT,
      quit: Code::QUIT,
      illegal_instruction: Code::ILLEGAL_INSTRUCTION,
      trace_trap: Code::TRACE_TRAP,
      abort: Code::ABORT,
      kill: Code::KILL,
      bus_error: Code::BUS_ERROR,
      memory_error: Code::MEMORY_ERROR,
      segmentation_fault: Code::MEMORY_ERROR,
      pipe: Code::PIPE,
      alarm: Code::ALARM,
      user1: Code::USER1,
      user2: Code::USER2
    }.freeze
    private_constant :NAME_TO_EXIT_CODE

    CODE_TO_EXIT_MESSAGE = {
      Code::SUCCESS => "Successful termination",
      Code::ERROR => "An error occurred",
      Code::SHELL_MISUSE => "Misuse of shell builtins",

      Code::USAGE_ERROR => "Command line usage error",
      Code::DATA_ERROR => "Data format error",
      Code::NO_INPUT => "Cannot open input",
      Code::NO_USER => "User name unknown",
      Code::NO_HOST => "Host name unknown",
      Code::SERVICE_UNAVAILABLE => "Service unavailable",
      Code::SOFTWARE_ERROR => "Internal software error",
      Code::SYSTEM_ERROR => "System error (e.g. can't fork)",
      Code::SYSTEM_FILE_MISSING => "Critical OS file missing",
      Code::CANT_CREATE => "Can't create user output file",
      Code::IO_ERROR => "Input/output error",
      Code::TEMP_FAIL => "Temp failure, user is invited to retry",
      Code::PROTOCOL => "Remote error in protocol",
      Code::NO_PERM => "Permission denied",
      Code::CONFIG_ERROR => "Configuration error",

      Code::CANNOT_EXECUTE => "Command invoked cannot execute",
      Code::COMMAND_NOT_FOUND => "Command not found",
      Code::INVALID_ARGUMENT => "Invalid argument",

      Code::HANGUP => "Hangup detected on controlling terminal or death of controlling process.",
      Code::INTERRUPT => "Interrupted by Control-C",
      Code::QUIT => "Quit program",
      Code::ILLEGAL_INSTRUCTION => "Illegal instruction",
      Code::TRACE_TRAP => "Trace/breakpoint trap",
      Code::ABORT => "Abort program",
      Code::KILL => "Kill program",
      Code::BUS_ERROR => "Access to an undefined portion of a memory object",
      Code::MEMORY_ERROR => "An invalid virtual memory reference or segmentation fault",
      Code::PIPE => "Write on a pipe with no one to read it",
      Code::ALARM => "Alarm clock",
      Code::USER1 => "User-defined signal 1",
      Code::USER2 => "User-defined signal 2"
    }.freeze
    private_constant :CODE_TO_EXIT_MESSAGE

    # Check if an exit code is valid, that it's within the 0-255 (inclusive)
    #
    # @param [Integer] code
    #   the code to check
    #
    # @return [Boolean]
    #
    # @api public
    def exit_valid?(code)
      code >= 0 && code <= 255
    end

    # Check if an exit code is already defined by Unix system
    #
    # @param [Integer] code
    #   the code to check
    #
    # @return [Boolean]
    #
    # @api public
    def exit_reserved?(code)
      (code >= Code::SUCCESS && code <= Code::SHELL_MISUSE) ||
        (code >= Code::USAGE_ERROR && code <= Code::CONFIG_ERROR) ||
        (code >= Code::CANNOT_EXECUTE && code <= Code::USER2)
    end

    # Check if the exit status was successful.
    #
    # @param [Integer] code
    #
    # @api public
    def exit_success?(code)
      code == Code::SUCCESS
    end

    # A user friendly explanation of the exit code
    #
    # @example
    #   TTY::Exit.exit_message(:usage_error)
    #   # => "Command line usage error"
    #
    # @param [String,Integer] name_or_code
    #
    # @api public
    def exit_message(name_or_code = :ok)
      (Registry.exits[name_or_code] || {})[:message] ||
        CODE_TO_EXIT_MESSAGE[exit_code(name_or_code)] || ""
    end

    # Provide a list of reserved status messages
    #
    # @api public
    def exit_messages
      CODE_TO_EXIT_MESSAGE
    end

    # Provide exit code for a name or status
    #
    # @example
    #   TTY::Exit.exit_code(:usage_error)
    #   # => 64
    #
    # @param [String,Integer] name_or_code
    #
    # @return [Integer]
    #   the exit code
    #
    # @api public
    def exit_code(name_or_code = :ok)
      case name_or_code
      when String, Symbol
        (Registry.exits[name_or_code.to_sym] || {})[:code] ||
          NAME_TO_EXIT_CODE.fetch(name_or_code.to_sym) do
            raise Error, "Name '#{name_or_code}' isn't recognized."
          end
      when Numeric
        if exit_valid?(name_or_code.to_i)
          name_or_code.to_i
        else
          raise Error, "Provided code outside of the range (0 - 255)"
        end
      else
        raise Error, "Provide a name or a number as an exit code"
      end
    end

    # Provide a list of reserved codes
    #
    # @api public
    def exit_codes
      NAME_TO_EXIT_CODE
    end

    # Exit this process with a given status code
    #
    # @param [String,Integer] name_or_code
    #   The name for an exit code or code itself
    # @param [String] message
    #   The message to print to io stream
    # @param [IO] io
    #   The io to print message to
    #
    # @return [nil]
    #
    # @api public
    def exit_with(name_or_code = :ok, message = nil, io: $stderr)
      if message == :default
        message = "ERROR(#{exit_code(name_or_code)}): #{exit_message(name_or_code)}"
      end
      io.print(message) if message
      ::Kernel.exit(exit_code(name_or_code))
    end
  end # Exit
end # TTY


lib/tty/exit/code.rb
# frozen_string_literal: true

module TTY
  module Exit
    # All reserved exit codes as constants
    module Code
      # No errors
      SUCCESS = 0

      # Catchall for general errors
      ERROR = 1

      # Missing keyword or command, or permission problem
      # (and diff return code on a failed binary file comparison).
      SHELL_MISUSE = 2

      # The start base code to reduce possibility of clashing with other
      # exit statuses that programs may already run.
      BASE = 64

      # The command was used incorrectly, e.g., with the wrong number
      # of arguments, a bad flag, a bad syntax in a parameter, or whatever.
      USAGE_ERROR = BASE # 64

      # The input data was incorrect in some way.
      # This should only be used for user's data & not system files.
      DATA_ERROR = BASE + 1 # 65

      # An input file (not a system file) did not exist or was not readable.
      NO_INPUT = BASE + 2 # 66

      # The user specified did not exist. This might be used for mail
      # addresses or remote logins.
      NO_USER = BASE + 3 # 67

      # The host specified did not exist. This is used in mail addresses
      # or network requests.
      NO_HOST = BASE + 4 # 68

      # A service is unavailable. This can occur if a support program
      # or file does not exist. This can also be used as a catchall
      # message when something you wanted to do doesn't work,
      # but you don't know why.
      SERVICE_UNAVAILABLE = BASE + 5 # 69

      # An internal software error has been detected. This should be
      # limited to non-operating system related errors as possible.
      SOFTWARE_ERROR = BASE + 6 # 70

      # An operating system error has been detected. This is intended
      # to be used for such things as "cannot fork", "cannot create pipe",
      # or the like. It includes things like getuid returning a user that
      # does not exist in the passwd file.
      SYSTEM_ERROR = BASE + 7 # 71

      # Some system file (e.g., /etc/passwd, /etc/utmp, etc.) does not exist,
      # cannot be opened, or has some sort of error (e.g., syntax error).
      SYSTEM_FILE_MISSING = BASE + 8 #72

      # A (user specified) output file cannot be created.
      CANT_CREATE = BASE + 9 # 73

      # An error occurred while doing I/O on some file.
      IO_ERROR = BASE + 10 # 74

      # Temporary failure, indicating something that is not really an error.
      # For example that a mailer could not create a connection, and the
      # request should be reattempted later.
      TEMP_FAIL = BASE + 11 # 75

      # The remote system returned something that was 'not possible' during
      # a protocol exchange.
      PROTOCOL = BASE + 12 # 76

      # You did not have sufficient permission to perform the operation.
      # This is not intended for file system problems, which should use
      # NO_INPUT or CANT_CREATE, but rather for higher level permissions.
      NO_PERM = BASE + 13 # 77

      # Something was found in an unconfigured or misconfigured state.
      CONFIG_ERROR = BASE + 14 # 78

      # Command invoked cannot execute. This may be due to permission
      # issues.
      CANNOT_EXECUTE = 126

      # "command not found", possible typos in shell command or
      # unrecognized characters
      COMMAND_NOT_FOUND = 127

      # The start base code for the system interrupt signals.
      SIGNAL_BASE = 128

      # Exit takes only integer args in the range 0 - 255,
      # e.g. exit 3.14159 is invalid.
      INVALID_ARGUMENT = SIGNAL_BASE

      # This indicates that program received SIGHUP signal.
      # It means that the controlling pseudo or virtual terminal
      # has been closed.
      HANGUP = SIGNAL_BASE + 1 # 129

      # This indicates that program received SIGINT signal.
      # An interrupt signal that by default this causes the
      # process to terminate.
      INTERRUPT = SIGNAL_BASE + 2 # 130

      # This indicates that program received SIGQUIT signal.
      QUIT = SIGNAL_BASE + 3 # 131

      # This indicates that program received SIGQUIT signal.
      ILLEGAL_INSTRUCTION = SIGNAL_BASE + 4 # 132

      # This indicates that program received SIGTRAP signal.
      TRACE_TRAP = SIGNAL_BASE + 5 # 133

      # This indicates that program received SIGABRT signal.
      ABORT = SIGNAL_BASE + 6 # 134

      # This indicates that program received SIGKILL signal.
      KILL = SIGNAL_BASE + 9 # 137

      # This indicates that program received SIGBUS signal.
      # Access to an undefined portion of a memory object
      BUS_ERROR = SIGNAL_BASE + 10 # 138

      # This indicates that program received SIGSEGV signal.
      MEMORY_ERROR = SIGNAL_BASE + 11 # 139

      # This indicates that program received SIGPIPE signal.
      PIPE = SIGNAL_BASE + 13 # 141

      # This indicates that program received SIGALARM signal.
      ALARM = SIGNAL_BASE + 14 # 142

      # This indicates that program received SIGUSR1 signal.
      USER1 = SIGNAL_BASE + 30 # 158

      # This indicates that program received SIGUSR2 signal.
      USER2 = SIGNAL_BASE + 31 # 159
    end # Code
  end # Exit
end # TTY
